tools/cfs-fuse.c

```diff
@@ -0,0 +1,1229 @@
+	first_block = buf.offset / EROFS_BLKSIZ;
+
+	if (first_block >= n_blocks) {
+		goto out;
+	}
+
+	/* First read the out-of-band blocks */
+	done = false;
+	for (uint64_t block = first_block; block < last_block; block++) {
+		size_t block_start = block * EROFS_BLKSIZ;
+		size_t block_size = EROFS_BLKSIZ;
+
+		if (!tailpacked && block + 1 == last_block) {
+			block_size = file_size % EROFS_BLKSIZ;
+			if (block_size == 0) {
+				block_size = EROFS_BLKSIZ;
+			}
+		}
+
+		if (buf.offset >= block_start &&
+		    buf.offset < block_start + block_size) {
+			const uint8_t *block_data =
+				erofs_data + raw_blkaddr * EROFS_BLKSIZ + block_start;
+			if (cfs_readdir_block(req, &buf, block_data, block_size,
+					      block_start, use_plus)) {
+				done = true;
+				break;
+			}
+		}
+	}
+
+	if (!done && tailpacked) {
+		size_t block_start = last_block * EROFS_BLKSIZ;
+		size_t block_size = file_size % EROFS_BLKSIZ;
+
+		if (buf.offset >= block_start &&
+		    buf.offset < block_start + block_size) {
+			const uint8_t *block_data =
+				((uint8_t *)cino) + isize + xattr_size;
+			cfs_readdir_block(req, &buf, block_data, block_size,
+					  block_start, use_plus);
+		}
+	}
+
+out:
+	fuse_reply_buf(req, (char *)bufdata, buf.current_size);
+}
+
+static void cfs_readdir(fuse_req_t req, fuse_ino_t ino, size_t max_size,
+			off_t off, struct fuse_file_info *fi)
+{
+	_cfs_readdir(req, ino, max_size, off, fi, false);
+}
+
+static void cfs_readdir_plus(fuse_req_t req, fuse_ino_t ino, size_t max_size,
+			     off_t off, struct fuse_file_info *fi)
+{
+	_cfs_readdir(req, ino, max_size, off, fi, true);
+}
+
+static void cfs_opendir(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi)
+{
+	const erofs_inode *cino = cfs_get_erofs_inode(ino);
+	mode_t mode;
+
+	if (cino == NULL) {
+		fuse_reply_err(req, ENOENT);
+		return;
+	}
+
+	mode = erofs_inode_get_mode(cino);
+
+	if ((mode & S_IFMT) != S_IFDIR) {
+		fuse_reply_err(req, ENOTDIR);
+		return;
+	}
+
+	fi->keep_cache = 1;
+	fi->cache_readdir = 1;
+	fi->fh = (uintptr_t)cino;
+	fuse_reply_open(req, fi);
+}
+
+static void cfs_readlink(fuse_req_t req, fuse_ino_t ino)
+{
+	const erofs_inode *cino = cfs_get_erofs_inode(ino);
+	uint32_t mode;
+	uint64_t file_size;
+	uint16_t xattr_icount;
+	uint32_t raw_blkaddr;
+	size_t isize;
+	size_t xattr_size;
+	bool tailpacked;
+	char name_buf[PATH_MAX];
+
+	if (cino == NULL) {
+		fuse_reply_err(req, ENOENT);
+		return;
+	}
+
+	erofs_inode_get_info(cino, &mode, &file_size, &xattr_icount,
+			     &raw_blkaddr, &isize);
+
+	if ((mode & S_IFMT) != S_IFLNK) {
+		fuse_reply_err(req, EINVAL);
+		return;
+	}
+
+	xattr_size = 0;
+	if (xattr_icount > 0)
+		xattr_size = sizeof(struct erofs_xattr_ibody_header) +
+			     (xattr_icount - 1) * 4;
+
+	tailpacked = erofs_inode_is_tailpacked(cino);
+	if (!tailpacked) {
+		fuse_reply_err(req, EINVAL);
+		return;
+	}
+
+	const uint8_t *inline_data = ((uint8_t *)cino) + isize + xattr_size;
+
+	memcpy(name_buf, inline_data, file_size);
```

Comment:
same here, do we need to make sure we don't write after the buffer?

Reply:
yeah, added